### DISCUSSION TITLE
How to Import a Block From a Different DXF File

### DISCUSSION THREAD

**Discussion by yuanshanhuhuan** [mozman/ezdxf#1221] [Q&A]
I am currently using Python 3.12.3 and the ezdxf library version 1.3.0 to manipulate DXF files. I am trying to insert a block named "防火锁" from the "9-114.dxf" file into the "9-118.dxf" file. The code I am using seems to run without throwing any exceptions. It successfully reads the source file, locates the block, and attempts to insert it into the target file. However, the block does not appear in the target "9-118.dxf" file after the insertion process. I can see from the print statements that the block is found in the source file and the insertion operation is executed, but when I open the target file, there is no visible trace of the inserted block.

Code Used:
```python
import ezdxf

def main():
    try:
        source_dxf = ezdxf.readfile("9-114.dxf")
        print("成功读取源文件 9-114.dxf")
        source_block_definitions = source_dxf.blocks
        print(f"源文件中获取到的块定义数量为: {len(source_block_definitions)}")
        if "防火锁" in source_block_definitions:
            block = source_block_definitions["防火锁"]
            print(f"在源文件中找到名为 '防火锁' 的块，块的一些属性示例（可按需调整查看内容）: {block.dxf.name}")
        else:
            print("在源文件 9-114.dxf 中未找到名为 '防火锁' 的块")
            return

        target_dxf = ezdxf.readfile("9-118.dxf")
        print("成功读取目标文件 9-118.dxf")
        target_msp = target_dxf.modelspace()
        target_msp.add_blockref("防火锁", (90, 1200), dxfattribs={'xscale': 1, 'yscale': 1, 'rotation': 0})
        print("已将 '防火锁' 块插入到目标文件 9-118.dxf 中")

        target_dxf.saveas("9-118.dxf")
        print("已成功保存修改后的 9-118.dxf 文件")
    except FileNotFoundError as e:
        print(f"未找到文件: {e.filename}")
    except ezdxf.DXFStructureError as e:
        print(f"文件结构错误: {e}")


if __name__ == "__main__":
    main()
```

Additional Information:
Python Version: 3.12.3
ezdxf Version: 1.3.0
The "9-114.dxf" and "9-118.dxf" files were created using [specify the software used to create them, if known].
I have already tried several troubleshooting steps, but none of them worked:
I have checked the layer visibility of both files, and all layers seem to be properly configured.
I have adjusted the block's scale, rotation, and position, but it still does not show up.
I have verified the integrity of the block definitions in both files, and they seem to be fine.
I have also checked the entity types within the block, and there does not seem to be anything unusual.
I have ensured that the files exist and are accessible.
I have tried using different DXF versions and simple test DXF files with basic blocks, but the problem persists.
Expected Result:
I expect the "防火锁" block from the "9-114.dxf" file to be inserted into the "9-118.dxf" file and be visible when I open the target file using a DXF viewer.
Actual Result:
The block appears to be inserted (as indicated by the print statements), but when I open the "9-118.dxf" file, there is no visible representation of the inserted block.
This issue has been causing me a lot of trouble, and I would appreciate any help or suggestions on how to resolve it. Thank you very much!

**Comment by mozman** [comment]
This is not a CAD application! There is no magic behind the scenes. Referencing a block from a different DXF file just don't work that way.

This example shows how to import a block from a different DXF file by the `ezdxf.xref` module. This works for simple files but I don't recommend such complex workflows because there will be use cases where this approach will fail:

```python
import ezdxf
from ezdxf import xref

SRC_BLK_NAME = "JustAnyBlockName"

def create_source_doc():
    doc = ezdxf.new()
    blk = doc.blocks.new(SRC_BLK_NAME)
    blk.add_circle((0, 0), radius=5)
    return doc

def main():
    sdoc = create_source_doc()
    tdoc = ezdxf.new()

    # import data from sdoc into tdoc
    loader = xref.Loader(sdoc, tdoc)
    blk = sdoc.blocks.get(SRC_BLK_NAME)
    # import block layout SRC_BLK_NAME
    loader.load_block_layout(blk)
    # run import process
    loader.execute()

    # test if SRC_BLK_NAME was imported
    assert SRC_BLK_NAME in tdoc.blocks
    msp = tdoc.modelspace()
    # create a new block reference to the imported  block
    msp.add_blockref(SRC_BLK_NAME, insert=(5, 5))
    
    tdoc.saveas("imported_block.dxf")


if __name__ == "__main__":
    main()

```

**Reply by yuanshanhuhuan** [reply]
Thank you very much for your help! I really appreciate that you took the time to give me suggestions. I am a beginner in Python programming and currently using the ezdxf library for DXF file operations. I have encountered some difficulties, but I have made some progress with your help.
My current goal is to copy the existing blocks from an existing source file (such as A.dxf) and insert them into a newly created B.dxf file. Moreover, I may need to perform such operations on multiple blocks, so it would be best to achieve batch processing. I am wondering if there are some good ways to complete this task?
What I understand so far is that for copying and inserting a single block, I can use some features of the ezdxf library as before. For example, I can use readfile to read the source file and the target file, then obtain block information through the blocks property, and then use xref.Loader to import the block layout from the source file to the target file. Finally, I can use add_blockref to add block references in the target file. However, when it comes to batch operations, I am a bit confused.
I am thinking, can I use a loop to traverse all the blocks in the source file and then perform the above copying and inserting operations on each block? But I am not sure if there will be any potential problems, such as performance issues or file structure confusion?
